README.md
# Smart_Calendar
## _Smart_Calendar_

Smart_Calendar is a library that show the calendar, with the possibility of selecting multiples days, change the month or year with simple dynamic.

## Features

- Predefined title widget that returns the calendar month and year.
- Customizable title widget.
- Possibility to change the month with horizontal sliding gesture.
- Brings an extra calendar for the months and year as a cafeteria, which can be reached by clicking on the title widget.
- The user can select multiple days.
- Returns the list of selected dates and returns the current month and year every time it changes.

## New Features
- Now it's possible to add a list of events dates with they title and description, check the example below to know how to do it.
- The list of events can be annual or not.

## Solved
- Solved the bug that was letting select the past dates.

## Attention
- Please use the example below of events dates to add events to your calendar, remember to add all parameters, the day and month are waiting for two digits, so, if the months is 1, then will be 01, the same for the day.

## Easy use
 - First you have to set the controller, where you have to set the initial and last date, the language, the calendar type and weekday type.
```sh
final controller = SmartCalendarController(
    initialDate: DateTime.now(),
    lastDate: DateTime.utc(2053, 04, 31),
    eventDates: [
      {
        "date": "2021-05-01",
        "description": "This a holiday because of the Worker day",
        "title": "Worker Day"
      },
      {
        "date": "2021-06-01",
        "description": "This a holiday because of the Kids day",
        "title": "Kids Day"
      },
      {
        "date": "2021-09-17",
        "description": "This a holiday because of the hero day",
        "title": "Hero day"
      }
    ],
    annualEvents: false,
    locale: 'en_US',
    calendarType: CalendarType.civilCalendar,
    weekdayType: WeekDayType.medium,
  );
```
 - After this you can call the class and add the functions that you want to return.
```sh
SmartCalendar(
                controller: controller,
                onBackwardOrForward: (month, year) {
                  setState(() {
                    this.month = month;
                    this.year = year;
                  });
                  print('This $month and this $year');
                },
                onDayAddedOrRemoved: (day, month, monthName, year, dates) {
                  print('Selected date: $year-$monthName-$day \n $dates');
                },
                )
```

## The Controller Parameters
- initialDate: This receive a DateTime that you want to be the initial date.

- lastDate: This receive a DateTime that you want to be the last date.

- eventDates: This receive a List of events dates with the parameters date, description and title.

- annualEvents: This receive a Boolean to say if the event list is annual or not.

- locale: This receive a String that will be locale for the intl get the exact translantion of thes month and weekdays.

- calendarType: This parementer can receive CalendarType.civilCalendar or CalendarType.notCivilCalendar. If you select civiCalendart, the week will start on Monday, but if you select notCivilCalendar, then will start with on a Sunday.
- 
- weekdayType: This paramenter can receive WeekDayType.short, WeekDayType.medium and WeekDayType.long. If you select short, then the weekday will show only the first letter, if you select medium, then the weekday will show only the 3 letters and if you select long, then the weekday will show the full name.

## SmartCalendar Parameters
- controller: This receive the controller that you initialized before.

- onBackwardOrForward: This receive a function so it may retrieve the month and year when gobackwar or forward on the month or year.

- onDayAddedOrRemoved: This receive a function so it may retrieve the day, month, monthName, year or list of date when a day is added or removed.

## Images of the library on use
 - This is smart Calendar with the default title widget
![Image 1](https://i.ibb.co/G9yYPsT/Whats-App-Image-2021-05-09-at-19-15-08.jpg)

 - This is Smart Calendar with the customizable title widget
![Image 2](https://i.ibb.co/FKMksbs/Whats-App-Image-2021-05-09-at-19-15-08-4.jpg)

 - This is smart Calendar with multiple days selected
![Image 3](https://i.ibb.co/ZYT04hr/Whats-App-Image-2021-05-09-at-19-15-08-2.jpg)

 - This is Smart Calendar with the extra year calendar showing
![Image 4](https://i.ibb.co/myj7pk5/Whats-App-Image-2021-05-09-at-19-15-08-3.jpg)

 - This is Smart Calendar with the extra month calendar showing
![Image 5](https://i.ibb.co/k8yzJhk/Whats-App-Image-2021-05-09-at-19-15-08-5.jpg)

 - This is Smart Calendar with calendartype as civilCalendar
![Image 6](https://i.ibb.co/YdK3zPh/Whats-App-Image-2021-05-09-at-19-15-08-7.jpg)

 - This is Smart Calendar with calendartype as notCivilCalendar
![Image 7](https://i.ibb.co/FKMksbs/Whats-App-Image-2021-05-09-at-19-15-08-4.jpg)

## Installation

Install the dependencies.

```sh
flutter pub get add smart_calendart
```

The library was completely made with dart and is still under development, hope this library can help you the same way is helping me. Here is the link to [Github](https://github.com/Marco4763/smart_calendar/blob/master/example/lib/main.dart) - Project example.

CHANGELOG.md
## [0.0.3] - TODO: 07-07-2021.

* TODO:
- The user will have the possibilty to enter the event on the app.
- Will be possible to change the color of the days.
- Parameter to say if will be multiple days or only one day selecting.


test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'file:///Users/mac/Documents/StudioProjects/smart_calendar/example/lib/main.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());

    // Verify that our counter starts at 0.
    expect(find.text('0'), findsOneWidget);
    expect(find.text('1'), findsNothing);

    // Tap the '+' icon and trigger a frame.
    await tester.tap(find.byIcon(Icons.add));
    await tester.pump();

    // Verify that our counter has incremented.
    expect(find.text('0'), findsNothing);
    expect(find.text('1'), findsOneWidget);
  });
}


lib/smart_calendar.dart
import 'package:flutter/material.dart';
import 'package:flutter_staggered_animations/flutter_staggered_animations.dart';
import 'package:get/get.dart';
import 'package:intl/date_symbol_data_local.dart';
import 'package:intl/intl.dart';
import 'package:smart_calendar/controller/smart_calendar_controller.dart';

class SmartCalendar extends StatefulWidget {
  SmartCalendar({
    @required this.controller,
    this.customTitleWidget,
    this.onBackwardOrForward,
    this.onDayAddedOrRemoved,
  });

  final SmartCalendarController controller;
  final Widget customTitleWidget;
  final Function(String month, int year) onBackwardOrForward;
  final Function(int day, int month, String monthName, int year, List dates)
      onDayAddedOrRemoved;

  @override
  _SmartCalendarState createState() => _SmartCalendarState();
}

class _SmartCalendarState extends State<SmartCalendar> {
  @override
  void initState() {
    super.initState();
    setWeekDaysAndMonths();
  }

  //This method will set the weekdays on the language that was inserted on the locale
  //and will order according to the type of the calendar inserted this method
  //is the first to run on the initialization of this class
  setWeekDaysAndMonths() {
    switch (widget.controller.calendarType) {
      case CalendarType.civilCalendar:
        initializeDateFormatting(widget.controller.locale);
        for (int i = 1; i < getWeekDays().length; i++) {
          widget.controller.daysOfTheWeek.add(getWeekDays()[i]);
        }
        widget.controller.daysOfTheWeek.add(getWeekDays()[0]);
        widget.controller.months = getMonths();
        break;

      case CalendarType.notCivilCalendar:
        initializeDateFormatting(widget.controller.locale);
        for (int i = 0; i < getWeekDays().length; i++) {
          widget.controller.daysOfTheWeek.add(getWeekDays()[i]);
        }
        widget.controller.months = getMonths();
        break;
    }
  }

  //This method get the weekdays according to the locale inserted
  //and is called on the setWeekDaysAndMonths() method
  List getWeekDays() {
    return DateFormat(
            '${widget.controller.initialDate}', widget.controller.locale)
        .dateSymbols
        .WEEKDAYS;
  }

  //This method get the months according to the locale inserted
  //and is called on the setWeekDaysAndMonths() method
  List getMonths() {
    return DateFormat(
            '${widget.controller.initialDate}', widget.controller.locale)
        .dateSymbols
        .MONTHS;
  }

  @override
  void dispose() {
    widget.controller.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Column(
      mainAxisSize: MainAxisSize.min,
      children: [_initCalendar()],
    );
  }

  //This method initialize the full calendar widget, set the default dates
  //and return the Calendar already build
  _initCalendar() {
    widget.controller.calcNumberOfDays();
    if (widget.controller.lastDate.year >= widget.controller.initialDate.year) {
      if (widget.controller.lastDate.year >
          widget.controller.initialDate.year) {
        if (widget.controller.weekDay != 0) {
          return _showCalendar();
        } else {
          return Container();
        }
      } else {
        if (widget.controller.lastDate.month >=
            widget.controller.initialDate.month) {
          if (widget.controller.weekDay != 0) {
            return _showCalendar();
          } else {
            return Container();
          }
        } else {
          return Container();
        }
      }
    } else {
      if (widget.controller.initialDate.month >=
          widget.controller.lastDate.month) {
        if (widget.controller.weekDay != 0) {
          return _showCalendar();
        } else {
          return Container();
        }
      } else {
        return Container();
      }
    }
  }

  //This method show the full calendar widget
  _showCalendar() {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.center,
      children: [
        GestureDetector(
          onTap: () {
            _buildMonthCalendar(context);
          },
          child: _buildCustomTitleWidget(),
        ),
        Row(
          mainAxisAlignment: MainAxisAlignment.spaceBetween,
          children: List.generate(
              widget.controller.daysOfTheWeek.length,
              (index) => Container(
                    width: 35,
                    height: 35,
                    child: Text(
                      _buildWeekDay(index),
                      textAlign: TextAlign.center,
                      style: TextStyle(
                          color: widget.controller.buildWeekendColor(index)),
                    ),
                  )),
        ),
        GestureDetector(
          onHorizontalDragEnd: (details) {
            if (details.primaryVelocity > 0) {
              setState(() {
                widget.controller
                    .goBackWard(function: widget.onBackwardOrForward);
              });
            } else if (details.primaryVelocity < 0) {
              setState(() {
                widget.controller
                    .goForWard(function: widget.onBackwardOrForward);
              });
            }
          },
          child: _buildCalendar(),
        )
      ],
    );
  }

  //This method build the month calendar on a snackBar
  //in case the user click on the title widget
  _buildMonthCalendar(BuildContext context) {
    var snackBar = SnackBar(
        duration: Duration(days: 1),
        content: Column(
          mainAxisSize: MainAxisSize.min,
          children: [
            Container(
              width: Get.width,
              height: 50,
              child: Row(
                mainAxisAlignment: MainAxisAlignment.spaceBetween,
                children: [
                  IconButton(
                      icon: Icon(
                        Icons.arrow_back_ios,
                        color: Colors.black,
                      ),
                      onPressed: () {
                        setState(() {
                          widget.controller.goBackWardInYear(
                              function: widget.onBackwardOrForward);
                        });
                      }),
                  GestureDetector(
                    onTap: () {
                      ScaffoldMessenger.of(context).removeCurrentSnackBar();
                      _buildYearCalendar(context);
                    },
                    child: Obx(() => Text('${widget.controller.currentYear}')),
                  ),
                  IconButton(
                      icon: Icon(
                        Icons.arrow_forward_ios,
                        color: Colors.black,
                      ),
                      onPressed: () {
                        setState(() {
                          widget.controller.goForWardInYear(
                              function: widget.onBackwardOrForward);
                        });
                      }),
                ],
              ),
            ),
            Container(
              width: Get.width,
              height: 200,
              child: GridView.builder(
                padding:
                    EdgeInsets.only(left: 5.0, right: 5.0, top: 10, bottom: 10),
                itemCount: widget.controller.months.length,
                gridDelegate: SliverGridDelegateWithFixedCrossAxisCount(
                    crossAxisCount: 3, childAspectRatio: 2.2),
                itemBuilder: (context, index) {
                  return GestureDetector(
                    onTap: () {
                      ScaffoldMessenger.of(context).removeCurrentSnackBar();
                      setState(() {
                        widget.controller.currentMonth = index + 1;
                        widget.controller.calcNumberOfDays(
                            function: widget.onBackwardOrForward);
                      });
                    },
                    child: Card(
                      color: _buildMonthCardColor(index),
                      shape: RoundedRectangleBorder(
                          borderRadius: BorderRadius.circular(80)),
                      child: Container(
                        width: 25,
                        height: 25,
                        child: Center(
                            child: Obx(() => Text(
                                  widget.controller.months[index]
                                      .toString()
                                      .capitalizeFirst,
                                  style: TextStyle(color: Colors.white),
                                ))),
                      ),
                    ),
                  );
                },
              ),
            ),
          ],
        ));
    ScaffoldMessenger.of(context).showSnackBar(snackBar);
  }

  //This method build the year calendar on a snackBar
  //in case the user click on the title widget
  _buildYearCalendar(BuildContext context) {
    var snackBar = SnackBar(
        duration: Duration(days: 1),
        content: Column(
          mainAxisSize: MainAxisSize.min,
          children: [
            Container(
              width: Get.width,
              height: 50,
              child: Center(
                child: Text(
                  'Calendário',
                  textAlign: TextAlign.center,
                  style: TextStyle(color: Colors.white),
                ),
              ),
            ),
            Container(
              width: Get.width,
              height: 200,
              child: GridView.builder(
                padding:
                    EdgeInsets.only(left: 5.0, right: 5.0, top: 10, bottom: 10),
                itemCount: widget.controller.getDifferenceBetweenYears(),
                gridDelegate: SliverGridDelegateWithFixedCrossAxisCount(
                    crossAxisCount: 3, childAspectRatio: 2.2),
                itemBuilder: (context, index) {
                  return GestureDetector(
                    onTap: () {
                      ScaffoldMessenger.of(context).removeCurrentSnackBar();
                      setState(() {
                        widget.controller.currentYear =
                            widget.controller.initialDate.year + index;
                        widget.controller.calcNumberOfDays(
                            function: widget.onBackwardOrForward);
                        _buildMonthCalendar(context);
                      });
                    },
                    child: Card(
                      color: _buildYearCardColor(index),
                      shape: RoundedRectangleBorder(
                          borderRadius: BorderRadius.circular(80)),
                      child: Container(
                        width: 25,
                        height: 25,
                        child: Center(
                            child: Text(
                          '${widget.controller.initialDate.year + index}',
                          style: TextStyle(color: Colors.white),
                        )),
                      ),
                    ),
                  );
                },
              ),
            ),
          ],
        ));
    ScaffoldMessenger.of(context).showSnackBar(snackBar);
  }

  //This method return the the color of the card depending
  //if the month is the current month or not
  _buildMonthCardColor(int index) {
    if ((widget.controller.currentMonth - 1) == index) {
      return Colors.red;
    } else {
      return Colors.grey;
    }
  }

  //This method return the the color of the card depending
  //if the year is the current year or not
  _buildYearCardColor(int index) {
    if (widget.controller.currentYear ==
        widget.controller.initialDate.year + index) {
      return Colors.red;
    } else {
      return Colors.grey;
    }
  }

  //This method build the title widget depending if there is
  //a custom one or will use the default title widget
  _buildCustomTitleWidget() {
    if (widget.customTitleWidget != null) {
      return widget.customTitleWidget;
    } else {
      return Container(
        width: Get.width,
        height: 50,
        child: Row(
          mainAxisAlignment: MainAxisAlignment.spaceBetween,
          children: [
            IconButton(
                icon: Icon(
                  Icons.arrow_back_ios,
                  color: Colors.black,
                ),
                onPressed: () {
                  setState(() {
                    widget.controller
                        .goBackWard(function: widget.onBackwardOrForward);
                  });
                }),
            Obx(() => Text(
                '${widget.controller.months[widget.controller.currentMonth - 1].toString().capitalizeFirst} - ${widget.controller.currentYear}')),
            IconButton(
                icon: Icon(
                  Icons.arrow_forward_ios,
                  color: Colors.black,
                ),
                onPressed: () {
                  setState(() {
                    widget.controller
                        .goForWard(function: widget.onBackwardOrForward);
                  });
                }),
          ],
        ),
      );
    }
  }

  //This method return the names of the each weekday, but first he check
  //the type of weekday he will return, short, medium or long
  _buildWeekDay(int index) {
    switch (widget.controller.weekdayType) {
      case WeekDayType.short:
        return widget.controller.daysOfTheWeek[index]
            .toString()
            .substring(0, 1)
            .capitalize;
        break;
      case WeekDayType.medium:
        return widget.controller.daysOfTheWeek[index]
            .toString()
            .substring(0, 3)
            .capitalizeFirst;
        break;
      case WeekDayType.long:
        return widget.controller.daysOfTheWeek[index]
            .toString()
            .capitalizeFirst;
        break;
    }
  }

  //This method build the calendar widget, the method return all the
  //days of the month and animate the way to show them
  _buildCalendar() {
    int numberOfWeeks = widget.controller
        .calcNumberOfWeeks(calendarType: widget.controller.calendarType);
    switch (widget.controller.calendarType) {
      case CalendarType.civilCalendar:
        return Column(
          children: List.generate(numberOfWeeks + 1, (index) {
            return AnimationConfiguration.staggeredList(
              position: index,
              child: SlideAnimation(
                verticalOffset: 50.0,
                child: FadeInAnimation(
                  child: _buildCalendarContent(index, numberOfWeeks),
                ),
              ),
            );
          }),
        );
        break;
      case CalendarType.notCivilCalendar:
        return Column(
          children: List.generate(numberOfWeeks, (index) {
            return AnimationConfiguration.staggeredList(
              position: index,
              child: SlideAnimation(
                verticalOffset: 50.0,
                child: FadeInAnimation(
                  child: _buildCalendarContent(index, numberOfWeeks),
                ),
              ),
            );
          }),
        );
        break;
    }
  }

  //This method build the calendar widget, the method check if the column position
  //is according to the number of weeks and set each day on each line
  _buildCalendarContent(int position, int numberOfWeeks) {
    if (position >= widget.controller.checkNumberWeekDay(numberOfWeeks)) {
      return Row(
        mainAxisAlignment: MainAxisAlignment.spaceBetween,
        children: List.generate(7, (index) {
          return _showDay(index, position, numberOfWeeks);
        }),
      );
    } else {
      return Container();
    }
  }

  //This will check the day and set it to each position according with he day of week, depending on
  //the type of calendar that was selected by the user when his was calling the smartCalendar
  dynamic _showDay(int index, position, numberOfWeeks) {
    switch (widget.controller.calendarType) {
      case CalendarType.civilCalendar:
        if (position == (numberOfWeeks - 5)) {
          if (index >= widget.controller.weekDay - 1) {
            int lastDay = 0;
            return _clickableText(
              position: position,
              index: index,
              day: widget.controller.numberOfDays[
                  lastDay + index - widget.controller.weekDay + 2],
            );
          } else {
            return Container(
              width: 35,
              height: 35,
            );
          }
        } else {
          int lastDay = 7;
          if (position == (numberOfWeeks - 5) + 1) {
            setState(() {
              lastDay = 7;
            });
          } else if (position == (numberOfWeeks - 5) + 2) {
            setState(() {
              lastDay = 14;
            });
          } else if (position == (numberOfWeeks - 5) + 3) {
            setState(() {
              lastDay = 21;
            });
          } else if (position == (numberOfWeeks - 5) + 4) {
            setState(() {
              lastDay = 28;
            });
          } else {
            setState(() {
              lastDay = 35;
            });
          }
          if (lastDay + index - widget.controller.weekDay + 2 >=
              widget.controller.numberOfDays.length) {
            return Container(
              width: 35,
              height: 35,
            );
          } else {
            return _clickableText(
              position: position,
              index: index,
              day: widget.controller.numberOfDays[
                      lastDay + index - widget.controller.weekDay + 1] +
                  1,
            );
          }
        }
        break;

      case CalendarType.notCivilCalendar:
        if (position == (numberOfWeeks - 6)) {
          if (index >= widget.controller.weekDay - 1) {
            int lastDay = 0;
            return _clickableText(
              position: position,
              index: index,
              day: widget.controller.numberOfDays[
                  lastDay + index - widget.controller.weekDay + 2],
            );
          } else {
            return Container(
              width: 35,
              height: 35,
            );
          }
        } else {
          int lastDay = 7;
          if (position == (numberOfWeeks - 6) + 1) {
            setState(() {
              lastDay = 7;
            });
          } else if (position == (numberOfWeeks - 6) + 2) {
            setState(() {
              lastDay = 14;
            });
          } else if (position == (numberOfWeeks - 6) + 3) {
            setState(() {
              lastDay = 21;
            });
          } else if (position == (numberOfWeeks - 6) + 4) {
            setState(() {
              lastDay = 28;
            });
          } else {
            setState(() {
              lastDay = 35;
            });
          }
          if (lastDay + index - widget.controller.weekDay + 2 >=
              widget.controller.numberOfDays.length) {
            return Container(
              width: 35,
              height: 35,
            );
          } else {
            return _clickableText(
              position: position,
              index: index,
              day: widget.controller.numberOfDays[
                      lastDay + index - widget.controller.weekDay + 1] +
                  1,
            );
          }
        }
        break;
    }
  }

  //This method build the text widget that show tha day, this method also handle the click
  //on each day so it can add or remove the selected day
  _clickableText({
    @required int position,
    @required int index,
    @required int day,
  }) {
    return Container(
        width: 35,
        height: 35,
        child: Obx(
          () => GestureDetector(
            onTap: () {
              setState(() {
                widget.controller.buildDayInset(
                    position, index, day, widget.onDayAddedOrRemoved, context);
              });
            },
            child: Card(
              elevation: 0,
              color: widget.controller.buildCardColor(position, index, day),
              shape: RoundedRectangleBorder(
                  borderRadius: BorderRadius.circular(80.0)),
              child: Center(
                child: widget.controller.checkEventDate(day, index),
              ),
            ),
          ),
        ));
  }
}


lib/widget/alert_calendar_widget.dart
import 'package:flutter/material.dart';
import 'package:get/get.dart';
import 'package:smart_calendar/controller/smart_calendar_controller.dart';

enum WidgetType { alertDialog, snackbar }

class MonthCalendarWidget extends StatefulWidget {
  const MonthCalendarWidget({
    Key key,
    @required this.widgetType,
    @required this.backWardClick,
    @required this.forWardClick,
    @required this.clickOverTitle,
    @required this.clickOnMonth,
  }) : super(key: key);
  final WidgetType widgetType;
  final GestureTapCallback backWardClick;
  final GestureTapCallback forWardClick;
  final GestureTapCallback clickOverTitle;
  final GestureTapCallback clickOnMonth;

  @override
  _MonthCalendarWidgetState createState() => _MonthCalendarWidgetState();
}

class _MonthCalendarWidgetState extends State<MonthCalendarWidget> {
  SmartCalendarController controller;

  @override
  void initState() {
    super.initState();
    if (Get.isRegistered<SmartCalendarController>()) {
      controller = Get.find<SmartCalendarController>();
    } else {
      controller = Get.put(SmartCalendarController());
    }
  }

  @override
  Widget build(BuildContext context) {
    return _buildMonthCalendarWidget();
  }

  _buildMonthCalendarWidget() {
    switch (widget.widgetType) {
      case WidgetType.alertDialog:
        return AlertDialog(
          contentPadding: EdgeInsets.all(0),
          content: Column(
            mainAxisSize: MainAxisSize.min,
            children: [
              Container(
                width: Get.width,
                height: 50,
                child: Row(
                  mainAxisAlignment: MainAxisAlignment.spaceBetween,
                  children: [
                    IconButton(
                      icon: Icon(
                        Icons.arrow_back_ios,
                        color: Colors.black,
                      ),
                      onPressed: widget.backWardClick,
                    ),
                    GestureDetector(
                      onTap: widget.clickOverTitle,
                      child: Obx(() => Text('${controller.currentYear}')),
                    ),
                    IconButton(
                      icon: Icon(
                        Icons.arrow_forward_ios,
                        color: Colors.black,
                      ),
                      onPressed: widget.forWardClick,
                    ),
                  ],
                ),
              ),
              Container(
                width: Get.width,
                height: 200,
                child: GridView.builder(
                  padding: EdgeInsets.only(
                      left: 5.0, right: 5.0, top: 10, bottom: 10),
                  itemCount: controller.months.length,
                  gridDelegate: SliverGridDelegateWithFixedCrossAxisCount(
                      crossAxisCount: 3, childAspectRatio: 2.2),
                  itemBuilder: (context, index) {
                    return GestureDetector(
                      onTap: widget.clickOnMonth,
                      child: Card(
                        color: _buildMonthCardColor(index),
                        shape: RoundedRectangleBorder(
                            borderRadius: BorderRadius.circular(80)),
                        child: Container(
                          width: 25,
                          height: 25,
                          child: Center(
                              child: Obx(() => Text(
                                    controller.months[index]
                                        .toString()
                                        .capitalizeFirst,
                                    style: TextStyle(color: Colors.white),
                                  ))),
                        ),
                      ),
                    );
                  },
                ),
              ),
            ],
          ),
        );
        break;

      case WidgetType.snackbar:
        return SnackBar(
            content: Column(
          mainAxisSize: MainAxisSize.min,
          children: [
            Container(
              width: Get.width,
              height: 50,
              child: Row(
                mainAxisAlignment: MainAxisAlignment.spaceBetween,
                children: [
                  IconButton(
                      icon: Icon(
                        Icons.arrow_back_ios,
                        color: Colors.black,
                      ),
                      onPressed: widget.backWardClick),
                  GestureDetector(
                    onTap: widget.clickOverTitle,
                    child: Obx(() => Text('${controller.currentYear}')),
                  ),
                  IconButton(
                      icon: Icon(
                        Icons.arrow_forward_ios,
                        color: Colors.black,
                      ),
                      onPressed: widget.forWardClick),
                ],
              ),
            ),
            Container(
              width: Get.width,
              height: 200,
              child: GridView.builder(
                padding:
                    EdgeInsets.only(left: 5.0, right: 5.0, top: 10, bottom: 10),
                itemCount: controller.months.length,
                gridDelegate: SliverGridDelegateWithFixedCrossAxisCount(
                    crossAxisCount: 3, childAspectRatio: 2.2),
                itemBuilder: (context, index) {
                  return GestureDetector(
                    onTap: widget.clickOnMonth,
                    child: Card(
                      color: _buildMonthCardColor(index),
                      shape: RoundedRectangleBorder(
                          borderRadius: BorderRadius.circular(80)),
                      child: Container(
                        width: 25,
                        height: 25,
                        child: Center(
                            child: Obx(() => Text(
                                  controller.months[index]
                                      .toString()
                                      .capitalizeFirst,
                                  style: TextStyle(color: Colors.white),
                                ))),
                      ),
                    ),
                  );
                },
              ),
            ),
          ],
        ));
        break;
    }
  }

  _buildMonthCardColor(int index) {
    if ((controller.currentMonth - 1) == index) {
      return Colors.red;
    } else {
      return Colors.grey;
    }
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:get/get.dart';
import 'package:smart_calendar/controller/smart_calendar_controller.dart';
import 'package:smart_calendar/smart_calendar.dart';

void main() => runApp(MyApp());

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  final controller = SmartCalendarController(
    initialDate: DateTime.now(),
    lastDate: DateTime.utc(2053, 04, 31),
    eventDates: [
      {
        "date": "2021-05-01",
        "description": "This a holiday because of the Worker day",
        "title": "Worker Day"
      },
      {
        "date": "2021-06-01",
        "description": "This a holiday because of the Kids day",
        "title": "Kids Day"
      },
      {
        "date": "2021-09-17",
        "description": "This a holiday because of the hero day",
        "title": "Hero day"
      }
    ],
    annualEvents: false,
    locale: 'en_US',
    calendarType: CalendarType.civilCalendar,
    weekdayType: WeekDayType.medium,
  );
  var month = 'Mês';
  var year = DateTime.now().year;

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      home: Scaffold(
        body: SafeArea(
            child: Container(
          width: Get.width / 1,
          child: Column(
            children: [
              SmartCalendar(
                controller: controller,
                onBackwardOrForward: (month, year) {
                  setState(() {
                    this.month = month;
                    this.year = year;
                  });
                  print('This $month and this $year');
                },
                onDayAddedOrRemoved: (day, month, monthName, year, dates) {
                  print('Selected date: $year-$monthName-$day \n $dates');
                },
              ),
            ],
          ),
        )),
      ),
    );
  }
}
